src/ray/raylet/worker_pool.cc

```diff
@@ -978,73 +982,65 @@ void WorkerPool::PopDeleteWorker(
-                                    worker,
-                                    PopWorkerStatus::OK,
-                                    &found,
-                                    &used);
-  RAY_LOG(DEBUG) << "PushWorker " << worker->WorkerId() << " used: " << used;
-  if (!used) {
-    // Put the worker to the idle pool.
+  {
+    auto it = std::find_if(
+        state.pending_registration_requests.begin(),
+        state.pending_registration_requests.end(),
+        [this, &worker](const std::shared_ptr<PopWorkerRequest> &pop_worker_request) {
+          return WorkerFitsForTask(*worker, *pop_worker_request) ==
+                 WorkerUnfitForTaskReason::NONE;
+        });
+    if (it != state.pending_registration_requests.end()) {
+      pop_worker_request = *it;
+      state.pending_registration_requests.erase(it);
+    }
+  }
+  if (!pop_worker_request) {
+    auto it = std::find_if(
+        state.pending_start_requests.begin(),
+        state.pending_start_requests.end(),
+        [this, &worker](const std::shared_ptr<PopWorkerRequest> &pop_worker_request) {
+          return WorkerFitsForTask(*worker, *pop_worker_request) ==
+                 WorkerUnfitForTaskReason::NONE;
+        });
+    if (it != state.pending_start_requests.end()) {
+      pop_worker_request = *it;
+      state.pending_start_requests.erase(it);
+    }
+  }
+
+  if (pop_worker_request) {
+    bool used = pop_worker_request->callback(worker, PopWorkerStatus::OK, "");
+    if (!used) {
+      // Retry PushWorker. Maybe it can be used by other tasks.
+      // Can we have tail call optimization for this? :)
+      return PushWorker(worker);
+    }
+  } else {
     state.idle.insert(worker);
     auto now = get_time_();
-    if (found) {
-      // If the worker was just started, then we should consider it first when
-      // choosing which idle workers to kill because it is cold.
+    if (worker->GetAssignedTaskTime() == absl::Time()) {
+      // If the worker is fresh started, then we should consider it first when
+      // choosing which idle workers to kill because it is not warmed up and is slower
+      // than those workers who served tasks before.
+      // See https://github.com/ray-project/ray/pull/36766
       idle_of_all_languages_.push_front(std::make_pair(worker, now));
     } else {
       idle_of_all_languages_.emplace_back(worker, now);
     }
-  } else if (!found) {
-    RAY_LOG(INFO) << "Worker not returned to the idle pool after being used. This may "
-                     "cause a worker leak, worker id:"
-                  << worker->WorkerId();
-  }
-  // We either have an idle worker or a slot to start a new worker.
-  if (worker->GetWorkerType() == rpc::WorkerType::WORKER) {
-    TryPendingPopWorkerRequests(worker->GetLanguage());
+    // We either have an idle worker or a slot to start a new worker.
+    if (worker->GetWorkerType() == rpc::WorkerType::WORKER) {
+      TryPendingPopWorkerRequests(worker->GetLanguage());
```

Comment:
```suggestion
      TryPendingStartPopWorkerRequests(worker->GetLanguage());
```

Reply:
You have two deque one is pending_registration and one is pending_start. This method is only trying pending_start deque?